### DISCUSSION TITLE
Changing flight modes

### DISCUSSION THREAD

**Discussion by evan-palmer** [Robotic-Decision-Making-Lab/blue#57] [Q&A]
Per @ssicari: After further investigating the use of QGroundControl (QGC), QGC recognizes the motors being armed and disarmed. It is not currently depicting the location of the vehicle, and when I try to run a basic way point mission it fails because it can not enter `GUIDED` mode. I have tried to manually test if I can put the vehicle into `GUIDED` mode via the ROS CLI but can not get it to recognize the command. Following the MAVROS docs, I can check the current mode it is in and see the possible modes it could change to, but everytime I try to actually change the mode, it says invalid passed service type. I was wondering if you have manually changed modes before and had a working command because the ones directly from the docs don't seem to work along with a few other variations I have tried. Thank you!

**Comment by evan-palmer** [comment]
Here is the command that you can use to set `GUIDED` mode on the vehicle: 

```bash
ros2 service call /mavros/set_mode mavros_msgs/srv/SetMode "{custom_mode: 'GUIDED'}"
```

**Comment by evan-palmer** [comment]
I would like to follow up with this to say though that, when I tried this out, I am able to set the mode to `GUIDED` mode, but it is still unable to execute any missions. I get the error 

```
Guided mode not supported by the Vehicle
```

I wonder if there might be some parameters that need to be modified to support this? I remember that reading some comments a while back that discussed something along the lines of GUIDED mode not being supported unless a GPS was available.

**Reply by ssicari** [reply]
Hmmm, I will look into that. Can I ask how you are checking to see if you entered guided mode? I am using this command 
"ros2 topic echo /mavros/state" and checking the mode field. After running the command you sent above it says that it was accepted and the send_mode response is true, but after echoing mavros/state it still says manual mode. I also checked in qgroundcontrol and while I still can not run missions because it failed to go into guided mode, I also can not change the modes manually to circle and position hold modes. I can change to all other modes, so the GPS could be the problem because I believe all 3 of those modes use GPS. As for the parameters, were do we adjust parameters to support GPS? Is this something we need to add to the model for a new sensor or are there config files we would put it in?

**Reply by ssicari** [reply]
I changed GPS from disabled to auto in the heavy ardusub.params and now the bluerov icon shows up in qgroundcontrol but it still says it failed to switch into guided mode

**Comment by evan-palmer** [comment]
@rakeshv24 do you have any ideas how to easily enable `GUIDED` mode? I remember you mentioning that you needed to do that for some of your docking work.

**Comment by evan-palmer** [comment]
@ssicari Here are are couple of related threads that might be useful:

- https://discuss.bluerobotics.com/t/unable-to-switch-to-guided-mode/14393/6
- https://discuss.bluerobotics.com/t/guided-mode-not-supported-by-vehicle/2993/3

These threads discuss that GPS feedback is required, so it may be necessary to integrate a GPS plugin to provide feedback to ArduSub from Gazebo.

**Reply by ssicari** [reply]
Ah I see, I will start working on that. Do you have any references of plugins you have done that I could look at to get a better sense of what making a plugin that interacts with gazebo and ardusub would entail? Thank you.

**Reply by ssicari** [reply]
I have a few more clarifying questions. So on the bluerov2_igniton project qgroundcontrol can enter guided mode, which leads me to believe it has a simulated gps. Do we know if this comes from the sitl simulated vehicle from ardusub. As well as, does the blue project start the sitl under the hood somewhere, or do we just not use it, if that makes sense. Thank you

**Comment by evan-palmer** [comment]
@ssicari I was chatting with @rakeshv24 and it sounds like the existing position information should be sufficient for enabling GUIDED mode, which indicates that there may be a bug elsewhere. The DVL integration uses the same interface [here](https://github.com/bluerobotics/BlueOS-Water-Linked-DVL/blob/7bb1bb5214922bdd546289ef9405f5a99ebb8113/dvl-a50/dvl.py#L236). I think we could try using the same parameters that the DVL uses to see if that works.

**Reply by rakeshv24** [reply]
Maybe try setting the EK3_GPS_TYPE to 3 like mentioned [here](https://github.com/bluerobotics/BlueOS-Water-Linked-DVL/blob/7bb1bb5214922bdd546289ef9405f5a99ebb8113/dvl-a50/dvl.py#L302).

**Reply by ssicari** [reply]
I see, for adding EK3_GPS_TYPE is that in the ardusubs.param or somewhere else? When I changed a previous parameter it was there. The list of parameters does not have the EK3_GPS_TYPE that I can find. I tried adding it myself but it still didn't go into guided mode so I am wondering if there is another place I need to update it. I will attach a picture below to try to illustrate what I mean. Thanks for taking the time to help, I appreciate it!


![image](https://github.com/evan-palmer/blue/assets/74316868/7eaf9096-6beb-4250-85f9-11ffd26a8d07)

**Reply by evan-palmer** [reply]
Yeah that would be the place to add it. It may be worth asking a question on the Blue Robotics discussion forums to see what their thoughts are. It seems like the parameters and state information needed for GUIDED mode are all being set/provided. My last guess would be that the position information isn't being accepted by ArduSub, though I would assume that if it isn't being provided correctly, there would be some feedback regarding that

**Reply by ssicari** [reply]
I looked at Clyde's Orca4 repo and managed to connect it to qgroundcontrol, his ardusub params list is the exact same as the screen shot I sent above. I am waiting on some responses from the ardupilot and bluerobotics forums for now, but have no idea what is going wrong to actually cause this. I may start creating and testing some ROS files in the mean time to see if I can get some kind of autonomy there by publishing to different thrusters without needing to switch into guided mode.

**Reply by evan-palmer** [reply]
Interesting are you able to switch into guided mode in Orca?

**Reply by ssicari** [reply]
Yeah, I can switch it into any mode and set way points to run a mission, and the gazebo model updates and carries out the mission. 
Only thing is I did get a warning for Bad NAV ALT 5000.00 but that might be unrelated and easy to fix, but it does allow me to switch to all the autonomous modes and carry out way point missions. 
Here is a link to the repo if you want to check it out: https://github.com/clydemcqueen/orca4/tree/main

**Reply by evan-palmer** [reply]
Ok. I'll look into the localizer today to see if that may be causing the problem. It's possible that the Gazebo localizer is sending the pose information in the wrong frame and ArduPilot is rejecting it which would cause this problem. I'm going to raise this to an issue as well.

**Reply by ssicari** [reply]
Alright sounds good, I can try to look at the localizer as well and also see if I can find any glaring differences between the blue repo and the orca repo to see why his works.
